crates/sui-core/src/authority.rs

```diff
@@ -2493,28 +2493,27 @@ impl AuthorityState {
     }
 
     pub fn get_object_read(&self, object_id: &ObjectID) -> SuiResult<ObjectRead> {
-        let Some(obj_ref) = self.database.get_object_or_tombstone(*object_id)? else {
-            return Ok(ObjectRead::NotExists(*object_id))
+        let Some((object_key, store_object)) = self.database.get_latest_object_or_tombstone(*object_id)? else {
+            return Ok(ObjectRead::NotExists(*object_id));
         };
-
-        if !obj_ref.2.is_alive() {
-            return Ok(ObjectRead::Deleted(obj_ref));
-        }
-
-        match self.read_object_at_version(object_id, obj_ref.1)? {
-            Some((object, layout)) => Ok(ObjectRead::Exists(obj_ref, object, layout)),
-            None => {
-                error!(
-                    "Object with in parent_entry is missing from object store, datastore is \
-                     inconsistent",
-                );
-                Err(UserInputError::ObjectNotFound {
-                    object_id: *object_id,
-                    version: Some(obj_ref.1),
-                }
-                .into())
-            }
-        }
+        if let Some(object_ref) = self
+            .database
+            .perpetual_tables
+            .tombstone_reference(&object_key, &store_object)?
+        {
+            return Ok(ObjectRead::Deleted(object_ref));
+        };
+        let object = self
+            .database
+            .perpetual_tables
+            .object(&object_key, store_object)?
+            .expect("Non tombstone store object could not be converted to object");
```

Comment:
is it preferable to panic here or return some `SuiError::UnexpectedError`?

Reply:
It should be a panic, as we already checked before that object is a not a tombstone before coming to this line